pkg/controller/deployment/util/deployment_util.go

```diff
@@ -699,6 +806,64 @@ func IsRollingUpdate(deployment *extensions.Deployment) bool {
 	return deployment.Spec.Strategy.Type == extensions.RollingUpdateDeploymentStrategyType
 }
 
+// IsDeploymentComplete considers a deployment to be complete once its desired replicas equals its
+// updatedReplicas and it doesn't violate minimum availability.
+func IsDeploymentComplete(deployment *extensions.Deployment, newStatus extensions.DeploymentStatus) bool {
+	return newStatus.UpdatedReplicas == deployment.Spec.Replicas &&
+		newStatus.AvailableReplicas >= deployment.Spec.Replicas-MaxUnavailable(*deployment)
+}
+
+// IsDeploymentProgressing reports progress for a deployment. Progress is estimated by comparing the
+// current with the new status of the deployment that the controller is observing. The following
+// algorithm is already used in the kubectl rolling updater to report lack of progress.
+func IsDeploymentProgressing(deployment *extensions.Deployment, newStatus extensions.DeploymentStatus) bool {
+	oldStatus := deployment.Status
+
+	// Old replicas that need to be scaled down
+	oldStatusOldReplicas := oldStatus.Replicas - oldStatus.UpdatedReplicas
+	newStatusOldReplicas := newStatus.Replicas - newStatus.UpdatedReplicas
+
+	return (newStatus.UpdatedReplicas > oldStatus.UpdatedReplicas) || (newStatusOldReplicas < oldStatusOldReplicas)
+}
+
+// used for unit testing
+var nowFn = func() time.Time { return time.Now() }
+
+// IsDeploymentFailed considers a deployment to be failed once its condition that reported progress
+// is older than the specified timeout.
+func IsDeploymentFailed(deployment *extensions.Deployment, newStatus extensions.DeploymentStatus) bool {
+	if deployment.Spec.ProgressDeadlineSeconds == nil {
+		return false
+	}
+
+	// Look at the difference in seconds between now and the last time we reported any
+	// progress or tried to create a replica set, or resumed a paused deployment and
+	// compare against progressDeadlineSeconds. If there is no such condition, we have
+	// no base to estimate progress.
+	var condition *extensions.DeploymentCondition
+out:
+	for i := range deployment.Status.Conditions {
+		cond := deployment.Status.Conditions[i]
+		if cond.Type != extensions.DeploymentProgressing {
+			continue
+		}
+		switch cond.Status {
+		case api.ConditionTrue, api.ConditionUnknown:
+			condition = &cond
+			break out
+		case api.ConditionFalse:
+			return true
+		}
+	}
+	if condition == nil {
+		return false
+	}
+
+	from := condition.LastTransitionTime
+	delta := time.Duration(*deployment.Spec.ProgressDeadlineSeconds) * time.Second
+	return from.Add(delta).Before(nowFn())
```

Comment:
This is one of the first places where we have perceived time used by the controller but recorded in the API server.  It's probably ok here, but clock skew during a controller failover will trigger a lack of progress storm.  I think we need to have an edge case check here where the controller, if it observes a progress condition in the future, at minimum warns or errors or logs *something*.  Please open a follow up issue to address that.